File: Generator.java
import processing.core.PApplet;

public abstract class Generator {

	PApplet p;
	String name;
	float amp;
	int phase;
	int period;

	public Generator(PApplet p, String name, float amp, int phase, int period) {
		this.name = name;
		this.p = p;
		this.amp = amp;
		this.phase = phase;
		this.period = period;
	}

	public String getName() {
		return name;
	}

	public void setAmp(float amp) {
		this.amp = amp;
	}

	public void setPhase(int phase) {
		this.phase = phase;
	}

	public void setPeriod(int period) {
		this.period = period;
	}

	public abstract float value(int millis);

}

File: python/parquette-lights/src/bellechasse/generators/generator.py
from typing import Optional
from abc import ABC, abstractmethod


class Generator(ABC):

    def __init__(
        self,
        *,
        name: Optional[str] = None,
        amp: float = 0.5,
        offset: float = 0.5,
        period: float = 1000,
        phase: float = 0
    ):
        self.name = name
        self.amp = amp
        self.phase = phase
        self.period = period
        self.offset = offset

    @abstractmethod
    def value(self, millis: float) -> float:
        pass
